server/module/services/utils/audio.py

```diff
@@ -218,53 +226,151 @@ def webrtc_vad_chunking(vad_level: int, chunk_size: int, raw_audio_bytes: bytes,
-    curr_end_secs = speech_timestamps[0]['end_secs']
+        speech_dict["start_secs"] = round(speech_dict["start"] / sample_rate, 3)
+        speech_dict["end_secs"] = round(speech_dict["end"] / sample_rate, 3)
+
+    adjusted_timestamps = adjust_timestamps(
+        speech_timestamps, sample_rate, max_chunk_duration_s, min_chunk_duration_s
+    )
+
+    audio_chunks: List[np.ndarray] = [
+        wav[timestamps["start"] : timestamps["end"]].numpy()
+        for timestamps in adjusted_timestamps
+    ]
+
+    return (audio_chunks, adjusted_timestamps)
+
+
+def adjust_timestamps(
+    speech_timestamps: List[Dict[str, float]],
+    sample_rate: int,
+    max_chunk_duration_s: float,
+    min_chunk_duration_s: float,
+):
+    """
+    Takes the speech timestamps output by the vad model and further
+    splits/merges based on the max_chunk_duration_s/min_chunk_duration_s.
+
+    Returns a list of adjusted timestamps.
+    """
+
+    adjusted_timestamps: List[Dict[str, float]] = []
+
+    curr_start = speech_timestamps[0]["start"]
+    curr_start_secs = speech_timestamps[0]["start_secs"]
+    curr_end_secs = speech_timestamps[0]["end_secs"]
+    chunk_duration = curr_end_secs - curr_start_secs
 
     for i in range(1, len(speech_timestamps)):
-        if min_chunk_duration_s and \
-            curr_end_secs - speech_timestamps[i]['start_secs'] < 3 \
-            and speech_timestamps[i]['end_secs'] - curr_start_secs <= min_chunk_duration_s:
+        chunk_gap = curr_end_secs - speech_timestamps[i]["start_secs"]
+        merged_chunks_duration = speech_timestamps[i]["end_secs"] - curr_start_secs
+        chunk_duration = curr_end_secs - curr_start_secs
+
+        # If the gap between the previous chunk and the current chunk is less
+        # than 3 seconds, and the previous chunk and current chunk put together
+        # are less than equal to the minimum chunk duration, merge the chunks
+        # by setting the end timestamp to the current chunk's end timestamp.
+        if chunk_gap < 3 and merged_chunks_duration <= min_chunk_duration_s:
+            curr_end_secs = speech_timestamps[i]["end_secs"]
+            continue
+
+        # Further pass the current chunk duration through windowed chunking to
+        # ensure that the size is within max_chunk_duration_s.y
+        chunked_timestamps = windowed_chunking(
+            curr_start,
+            curr_start_secs,
+            chunk_duration,
+            max_chunk_duration_s,
+            sample_rate,
+        )
+        adjusted_timestamps.extend(chunked_timestamps)
+
+        curr_start = speech_timestamps[i]["start"]
+        curr_start_secs = speech_timestamps[i]["start_secs"]
+        curr_end_secs = speech_timestamps[i]["end_secs"]
+
+    # One last chunk will always be left, add it to the adjusted_timestamps
+    chunked_timestamps = windowed_chunking(
+        curr_start,
+        curr_start_secs,
+        chunk_duration,
+        max_chunk_duration_s,
```

Comment:
Hey the `chunk_duration` should be recomputed na?
Currently it is using the old value.

Reply:
Damn nice catch. Fixed!